### PR TITLE
Fix Balance Sheet & Income Tax comparisons report by dates or by periods

### PR DESCRIPTION
**Pull request by ylavoie**
Restore Balance Sheet & Income Tax Comparison reports when selection is done by dates or by periods

### CODE REVIEW

File: lib/LedgerSMB/Scripts/pnl.pm [hunk 1]
```diff
@@ -74,9 +74,8 @@ sub generate_income_statement {
         $rpt = LedgerSMB::Report::PNL::Income_Statement->new(
             %$request,
             formatter_options => $request->formatter_options,
-            from_date  => $request->parse_date( $request->{from_date} ),
-            to_date  => $request->parse_date( $request->{to_date} ),
             column_path_prefix => [ 0 ]);
+        #die np $rpt;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/pnl.pm:78]: could you remove this stray comment?

**Comment by ehuelsmann** [comment]
I'm afraid that fixing the comparisons broke the date intervals...

**Comment by ehuelsmann** [comment]
@ylavoie how about approaching this differently: how about removing the complexity of needing to resolve this all server-side? How about making a bit of JS which calculates the correct start and end dates and sets the from and to dates accordingly?

File: lib/LedgerSMB/Scripts/pnl.pm (hunk 1 above)
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/pnl.pm]: Done

File: lib/LedgerSMB/Report/Dates.pm [hunk 2]
```diff
@@ -150,9 +149,9 @@ sub _collect_dates_comparisons {
 
     foreach my $i (1 .. $args{comparison_periods}) {
         push @dates, {
-            from_date => LedgerSMB::PGDate->from_input($args{"from_date_$i"}),
-            to_date => LedgerSMB::PGDate->from_input($args{"to_date_$i"}),
-            column_path_prefix => [ $i ]
+            from_date => LedgerSMB::PGDate->from_input($args{"from_date_$i"},$args{formatter_options}),
+            to_date => LedgerSMB::PGDate->from_input($args{"to_date_$i"},$args{formatter_options}),
+            column_path_prefix => [ sprintf('%02u', $i) ]
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Report/Dates.pm:154]: Nice find! I guess I never tested with more than 9 columns...

File: lib/LedgerSMB/Scripts/reports.pm [hunk 3]
```diff
@@ -193,27 +193,25 @@ Generates a balance sheet
...
-        to_date  => $request->parse_date( $request->{to_date} ),
+        from_date  => $request->{from_date} ? $request->parse_date( $request->{from_date} ) : undef,
+        to_date  => $request->{to_date} ? $request->parse_date( $request->{to_date} ) : undef,
         column_path_prefix => [ 0 ]);
     $rpt->run_report($request);
 
-    for my $key (qw(from_month from_year from_date to_date internal)) {
-        delete $request->{$_} for (grep { /^$key/ } keys %$request);
+    for my $key (qw(from_date to_date date_from date_to)) {
+        delete $rpt->{$_} for (grep { /^$key/ } keys %$rpt);
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/reports.pm:207]: why do we not need to delete `from_month` and `from_year`?

Comment (ylavoie) [reply lib/LedgerSMB/Scripts/reports.pm]: Once builders ran, all comparisons are set and we don't even need to cleanup $request or $rpt between passes.

**Pull request merged**